portal-impl/src/com/liferay/portlet/internal/PortletURLImpl.java

```diff
@@ -960,21 +1003,125 @@ protected String generateToString() {
 			}
 		}
 
-		Map<String, String[]> renderParams = _params;
+		Map<String, String[]> portletURLParams = new LinkedHashMap<>();
+
+		Set<String> actionParameterNames;
+
+		if (_mutableActionParameters != null) {
```

Comment:
@shuyangzhou I recommend you continue your code review by studying the changes I made to the `generateToString()` method. As you can see from here at line 1010, the `generateToString()` method was refactored so that it generates String representations of PortletURLs by honoring the distinction between action parameters, render parameters, and resource parameters.

The guiding principle that I used when refactoring the `generateToString()` method was to have it be 100% backward compatible. In other words, the return value of `PortletURL.toString()` had to be exactly the same before and after the refactor.